### PR TITLE
io fixes on non-blocking FDs

### PR DESCRIPTION
**Pull request by rjarry**
These are fixes for problems I encountered with the UNIX transport. Basically, when sending large configurations to netopeer2-server, since the socket fd is non-blocking, we can get `EAGAIN` errors upon `write()` if the socket buffer is full and the server does not read fast enough.

These errors are already handled in libnetconf2 but errno is not reset and libyang checks it (since commit https://github.com/CESNET/libyang/commit/609d708ca567d5400f303e654e28637203e1d07e). This causes the operation to fail with the following error:

```
Print error (Resource temporarily unavailable).
```

Work around the problem by resetting `errno` to `0` in case of `EAGAIN` or `EINTR`.

Also, if the error is `EINTR` do not wait before trying to read/write again.

I am not sure this is the correct approach, I proposed another (maybe complementary?) solution in https://github.com/CESNET/libyang/pull/903

### CODE REVIEW

**Comment by brchiu** [comment]
@rjarry  
Thanks for addressing this issue. I also encountered a similar one for SSH case with `Print error (No such file or directory)` error message. Can your approach apply to it ?

File: src/io.c [hunk 1]
```diff
@@ -190,7 +197,8 @@ nc_read(struct nc_session *session, char *buf, size_t count, uint32_t inact_time
 
         if (r == 0) {
             /* nothing read */
-            usleep(NC_TIMEOUT_STEP);
+            if (!interrupted)
+                usleep(NC_TIMEOUT_STEP);
```
Comment (michalvasko) [inline src/io.c]: Please always use `{` and `}`.

**Comment by michalvasko** [comment]
Hi,
depending on how we resolve cesnet/libyang#903, resetting `errno` may not be needed.

Regards,
Michal

**Comment by rjarry** [comment]
> @rjarry
> Thanks for addressing this issue. I also encountered a similar one for SSH case with `Print error (No such file or directory)` error message. Can your approach apply to it ?

@brchiu I don't think so, I only handle `EAGAIN` and `EINTR`, `No such file or directory` is `ENOENT` and is a real error.

File: src/io.c (hunk 1 above)
Comment (rjarry) [reply src/io.c]: done.

**Comment by rjarry** [comment]
> depending on how we resolve [CESNET/libyang#903](https://github.com/CESNET/libyang/pull/903), resetting `errno` may not be needed.

Indeed, I removed the first commit.

**Comment by michalvasko** [comment]
Hi,
it is fine now, thanks.

Regards,
Michal

**Comment by rjarry** [comment]
:+1:

**Pull request merged**